src/site/content/en/blog/2d-graphics-in-emscripten/index.md

```diff
@@ -0,0 +1,509 @@
+  SDL_Renderer *renderer;
+  SDL_CreateWindowAndRenderer(300, 300, 0, &window, &renderer);
+
+  // Set a color for drawing matching the earlier `ctx.fillStyle = "green"`.
+  SDL_SetRenderDrawColor(renderer, /* RGBA */ 0x00, 0x80, 0x00, 0xFF);
+  // Create and draw a rectangle like in the earlier `ctx.fillRect()`.
+  SDL_Rect rect = {.x = 10, .y = 10, .w = 150, .h = 100};
+  SDL_RenderFillRect(renderer, &rect);
+
+  // Render everything from a buffer to the actual screen.
+  SDL_RenderPresent(renderer);
+
+  // TODO: cleanup
+}
+```
+
+When linking with Emscripten, you need to use `-s USE_SDL=2`. This will tell Emscripten to fetch the SDL2 library, already precompiled to WebAssembly, and link it with your main application.
+
+```bash
+emcc example.cpp -o example.html -s USE_SDL=2
+```
+
+When the example is loaded in the browser, you'll see the familiar green rectangle:
+
+{% Img src="image/9oK23mr86lhFOwKaoYZ4EySNFp02/34ztXnZ2OZvWLwYdBzJd.png", alt="Emscripten-generated HTML page showing a green rectangle on a black square canvas", width="800", height="581" %}
+
+This code has a couple of problems though. First, it lacks proper cleanup of allocated resources. Second, on the web, pages don't get closed automatically when an application has finished its execution, so the image on the canvas gets preserved. However, when the same code is recompiled natively with
+
+```bash
+clang example.cpp -o example -lSDL2
+```
+
+and executed, the created window will only blink briefly and immediately close upon exit, so the user doesn't have a chance to see the image.
+
+### Integrating an event loop
+
+A more complete and idiomatic example would look need to wait in an event loop until the user chooses to quit the application:
+
+```cpp/14-20
+#include <SDL2/SDL.h>
+
+int main() {
+  SDL_Init(SDL_INIT_VIDEO);
+
+  SDL_Window *window;
+  SDL_Renderer *renderer;
+  SDL_CreateWindowAndRenderer(300, 300, 0, &window, &renderer);
+
+  SDL_SetRenderDrawColor(renderer, /* RGBA */ 0x00, 0x80, 0x00, 0xFF);
+  SDL_Rect rect = {.x = 10, .y = 10, .w = 150, .h = 100};
+  SDL_RenderFillRect(renderer, &rect);
+  SDL_RenderPresent(renderer);
+
+  while (1) {
+    SDL_Event event;
+    SDL_PollEvent(&event);
+    if (event.type == SDL_QUIT) {
+      break;
+    }
+  }
+
+  SDL_DestroyRenderer(renderer);
+  SDL_DestroyWindow(window);
+
+  SDL_Quit();
+}
+```
+
+After the image has been drawn to a window, the application now waits in a loop, where it can process keyboard, mouse and other user events. When the user closes the window, they'll trigger an `SDL_QUIT` event, which will be intercepted to exit the loop. After the loop is exited, the application will do the cleanup and then exit itself.
+
+Now compiling this example on Linux works as expected and shows a 300 by 300 window with a green rectangle:
+
+{% Img src="image/9oK23mr86lhFOwKaoYZ4EySNFp02/nUxwRMyWVmdK5zy2Kct6.png", alt="A square Linux window with black background and a green rectangle", width="400", height="440" %}
```

Comment:
This is weird, I set it to only 400x440, but it renders as a huge full-width image.
@rachelandrew any ideas on how to shrink it?